source/Ship.cpp

```diff
@@ -3498,20 +3508,56 @@ const vector<weak_ptr<Ship>> &Ship::GetEscorts() const
 // cues and try to stay with it when it lands or goes into hyperspace.
 void Ship::AddEscort(Ship &ship)
 {
+	RegisterEscort(ship);
 	escorts.push_back(ship.shared_from_this());
 }
 
 
 
 void Ship::RemoveEscort(const Ship &ship)
 {
+	// Reset this value, we will re-scan the list of escorts to set the
+	// escorts velocity based on remaining active escorts.
+	escortsVelocity = -1.;
+
 	auto it = escorts.begin();
-	for( ; it != escorts.end(); ++it)
-		if(it->lock().get() == &ship)
-		{
-			escorts.erase(it);
-			return;
+	while(it != escorts.end())
+	{
+		auto escort = it->lock().get();
+		if(escort){
+			if(escort == &ship){
+				escorts.erase(it);
+				continue;
+			}
+			else
+				// Re-scan existing escorts to cache data like the
+				// escorts velocity.
+				RegisterEscort(ship);
 		}
+		++it;
+	}
+}
+
+
+
+// Store relevant cached data for the given escort.
+void Ship::RegisterEscort(const Ship &ship)
```

Comment:
I'm not sold on the name, but 🤷‍♂️

Reply:
I agree the name is not great, but I didn't have inspiration yet to come up with something better.